File: src/main/java/komu/blunt/ast/ASTLambda.java
package komu.blunt.ast;

import komu.blunt.objects.Symbol;

import static com.google.common.base.Preconditions.checkNotNull;

public final class ASTLambda extends ASTExpression {
    public final Symbol argument;
    public final ASTExpression body;

    ASTLambda(Symbol argument, ASTExpression body) {
        this.argument = checkNotNull(argument);
        this.body = checkNotNull(body);
    }

    @Override
    public String toString() {
        return "(lambda " + argument + " " + body + ")";
    }

    @Override
    public ASTExpression simplify() {
        return new ASTLambda(argument, body.simplify());
    }
}

File: src/main/kotlin/komu/blunt/ast/ASTLambda.kt
package komu.blunt.ast

import komu.blunt.objects.Symbol

class ASTLambda(val argument: Symbol, val body: ASTExpression) : ASTExpression() {
    override fun toString() = "(lambda $argument $body)"
    override fun simplify() = ASTLambda(argument, body.simplify())
}
